jobs/build-charms/charms.py

```diff
@@ -39,46 +41,299 @@
+            self._echo("No resources for {}".format(charm_id))
+        resources_args = [
+            (
+                "--resource",
+                "{}-{}".format(resource["name"], resource["revision"]),
+            )
+            for resource in charm_resources
+        ]
+        sh.charm.release(
+            charm_id, f"--channel={to_channel}", *resources_args, _out=self._echo
+        )
+
+    def grant(self, charm_id):
+        sh.charm.grant(charm_id, "everyone", acl="read", _out=self._echo)
+
+    def upload(self, dst_path, entity):
+        out = sh.charm.push(dst_path, entity, _out=self._echo)
+        self._echo(f"Charm push returned: {out}")
+
+        # Output includes lots of ansi escape sequences from the docker push,
+        # and we only care about the first line, which contains the url as yaml.
+        out = yaml.safe_load(out.stdout.decode().strip().splitlines()[0])
+        return out["url"]
+
+    def set(self, entity, commit):
+        self._echo(f"Setting {entity} metadata: {commit}")
+        sh.charm.set(entity, f"commit={commit}", _out=self._echo)
+
+    def upload_resource(self, entity, resource_name, resource):
+        # If the resource is a file, populate the path where it was built.
+        # If it's a custom image, it will be in Docker and this will be a no-op.
+        retry_call(
+            sh.charm.attach,
+            fargs=[
+                entity,
+                f"{resource_name}={resource}",
+            ],
+            fkwargs={"_out": self._echo},
+            delay=2,
+            backoff=2,
+            tries=15,
+            exceptions=sh.ErrorReturnCode,
         )
 
+
+class _CharmHub:
+    HEADER_RE = re.compile(r"\W+\s{2}")
+    STATUS_RESOURCE = re.compile(r"(\S+) \(r(\d+)\)")
+
+    def __init__(self, build_entity):
+        self._echo = build_entity.echo
+
+    def _table_to_list(self, header, body):
+        if not body:
+            return None
+        rows = []
+        for line in body:
+            row, head = {}, line
+            for key in reversed(self.HEADER_RE.split(header)):
+                head, *value = head.strip().rsplit("  ", 1)
+                head, value = (head, value[0]) if value else ("", head.strip())
+                row[key] = value or rows[-1].get(key)
+            rows.append(row)
+        return rows
+
+    @staticmethod
+    def refresh(name, channel=None, architecture=None, base_channel=None):
+        channel = channel or "stable"
+        architecture = architecture or "amd64"
+        base_channel = base_channel or "20.04"
+        data = {
+            "context": [],
+            "actions": [
+                {
+                    "name": name,
+                    "base": {
+                        "name": "ubuntu",
+                        "architecture": architecture,
+                        "channel": base_channel,
+                    },
+                    "channel": channel,
+                    "action": "install",
```

Comment:
```suggestion
                    "action": "refresh",
```
Should this be refresh here?

Reply:
i dunno?  do you have some evidence?  I can be convinced.

Reply:
After digging into the [docs](https://api.snapcraft.io/docs/refresh.html) a bit they seem to suggest using the "download" action, which is the same as the install action, but is reserved for when you don't actually install anything. It does not really say why this distinction needs to be made however ¯\_(ツ)_/¯